diff --git a/CMakeLists.txt b/CMakeLists.txt
index 9f81b8bb..f1b05245 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -39,8 +39,6 @@ target_include_directories(${PROJECT_NAME} ${INCLUDE_SPECIFIER} INTERFACE
 
 target_link_libraries(${PROJECT_NAME} INTERFACE)
 
-# only install if top level project
-if(${CMAKE_PROJECT_NAME} STREQUAL ${PROJECT_NAME})
     # Steps here based on excellent guide: https://dominikberner.ch/cmake-interface-lib/
     # Which also details all steps
     include(CMakePackageConfigHelpers)
@@ -74,7 +72,6 @@ if(${CMAKE_PROJECT_NAME} STREQUAL ${PROJECT_NAME})
         DESTINATION ${CMAKE_INSTALL_DATAROOTDIR}/${PROJECT_NAME}/cmake)
     install(DIRECTORY ${PROJECT_SOURCE_DIR}/include/etl DESTINATION include)
 
-endif()
 
 if (BUILD_TESTS)
     enable_testing()
