diff --git a/CMakeLists.txt b/CMakeLists.txt
index cbe4564d..2fa1b994 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -39,8 +39,10 @@ target_include_directories(${PROJECT_NAME} ${INCLUDE_SPECIFIER} INTERFACE
 
 target_link_libraries(${PROJECT_NAME} INTERFACE)
 
+option(ENABLE_INSTALL "Enable installation" ON)
+
 # only install if top level project
-if(${CMAKE_PROJECT_NAME} STREQUAL ${PROJECT_NAME})
+if(ENABLE_INSTALL)
     # Steps here based on excellent guide: https://dominikberner.ch/cmake-interface-lib/
     # Which also details all steps
     include(CMakePackageConfigHelpers)
