diff --git a/CMakeLists.txt b/CMakeLists.txt
new file mode 100644
index 0000000..5d324ea
--- /dev/null
+++ b/CMakeLists.txt
@@ -0,0 +1,64 @@
+cmake_minimum_required(VERSION 3.16.0)
+
+include(CMakePackageConfigHelpers)
+include(GNUInstallDirs)
+
+list(APPEND libbcryp_srcs
+            bcrypt.c)
+
+list(APPEND crypt_blowfish_srcs
+            crypt_blowfish/crypt_blowfish.c
+            crypt_blowfish/crypt_gensalt.c
+            crypt_blowfish/wrapper.c)
+
+project(bcrypt VERSION 1.0.0 LANGUAGES CXX)
+
+add_library(${PROJECT_NAME} STATIC ${libbcryp_srcs}
+                        ${crypt_blowfish_srcs})
+
+if (CMAKE_SYSTEM_NAME MATCHES ".*BSD.*" )
+  target_compile_definitions(${PROJECT_NAME} PRIVATE "__SKIP_GNU")
+endif()
+
+set_target_properties(${PROJECT_NAME}
+    PROPERTIES
+        C_STANDARD          17
+        C_STANDARD_REQUIRED ON
+        PUBLIC_HEADER       "${${PROJECT_NAME}_PUBLIC_HEADERS}"
+        EXPORT_NAME         framework
+)
+
+target_compile_options(${PROJECT_NAME}
+    PRIVATE
+        # GCC
+        $<$<C_COMPILER_ID:GNU>:
+            -Wno-cpp
+        >
+)
+
+target_include_directories(${PROJECT_NAME}
+    PUBLIC
+        $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>
+        $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/${PROJECT_NAME}>
+)
+
+
+option(ENABLE_INSTALL "Enable installation of the library" Off)
+
+if (ENABLE_INSTALL)
+  ## Target installation
+  install(TARGETS   ${PROJECT_NAME}
+      EXPORT        ${PROJECT_NAME}Targets
+      ARCHIVE       DESTINATION ${CMAKE_INSTALL_LIBDIR}
+      LIBRARY       DESTINATION ${CMAKE_INSTALL_LIBDIR}
+      PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
+  )
+
+  install(FILES bcrypt.h DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
+
+  ## Target's cmake files: targets export
+  install(EXPORT  ${PROJECT_NAME}Targets
+      NAMESPACE   ${PROJECT_NAME}::
+      DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/${PROJECT_NAME}
+  )
+endif()
diff --git a/bcrypt.c b/bcrypt.c
index bd8722b..ac051c5 100644
--- a/bcrypt.c
+++ b/bcrypt.c
@@ -14,15 +14,32 @@
 #include <string.h>
 #include <sys/types.h>
 #include <sys/stat.h>
+#if (defined(__linux__) || defined(__unix__) || defined(__APPLE__) || defined(__FreeBSD__))
 #include <fcntl.h>
 #include <unistd.h>
 #include <errno.h>
+#endif
+
+#if (defined(CONFIG_IDF_TARGET_ESP32) || \
+     defined(ESP8266) || \
+     defined(CONFIG_IDF_TARGET_ESP32S2) || \
+     defined(CONFIG_IDF_TARGET_ESP32S3) || \
+     defined(CONFIG_IDF_TARGET_ESP32C2) || \
+     defined(CONFIG_IDF_TARGET_ESP32C3) || \
+     defined(CONFIG_IDF_TARGET_ESP32C5) || \
+     defined(CONFIG_IDF_TARGET_ESP32C6) || \
+     defined(CONFIG_IDF_TARGET_ESP32H2) || \
+     defined(CONFIG_IDF_TARGET_ESP32P4))
+#include <esp_random.h>
+#endif
 
 #include "bcrypt.h"
 #include "crypt_blowfish/ow-crypt.h"
+#include "crypt_blowfish/crypt_gensalt.h"
 
 #define RANDBYTES (16)
 
+#if (defined(__linux__) || defined(__unix__) || defined(__APPLE__) || defined(__FreeBSD__))
 static int try_close(int fd)
 {
 	int ret;
@@ -60,6 +77,7 @@ static int try_read(int fd, char *out, size_t count)
 
 	return 0;
 }
+#endif
 
 /*
  * This is a best effort implementation. Nothing prevents a compiler from
@@ -96,11 +114,12 @@ static int timing_safe_strcmp(const char *str1, const char *str2)
 
 int bcrypt_gensalt(int factor, char salt[BCRYPT_HASHSIZE])
 {
-	int fd;
 	char input[RANDBYTES];
 	int workf;
 	char *aux;
 
+#if (defined(__linux__) || defined(__unix__) || defined(__APPLE__) || defined(__FreeBSD__))
+	int fd;
 	fd = open("/dev/urandom", O_RDONLY);
 	if (fd == -1)
 		return 1;
@@ -114,6 +133,19 @@ int bcrypt_gensalt(int factor, char salt[BCRYPT_HASHSIZE])
 	if (try_close(fd) != 0)
 		return 3;
 
+#elif (defined(CONFIG_IDF_TARGET_ESP32) || \
+       defined(ESP8266) || \
+       defined(CONFIG_IDF_TARGET_ESP32S2) || \
+       defined(CONFIG_IDF_TARGET_ESP32S3) || \
+       defined(CONFIG_IDF_TARGET_ESP32C2) || \
+       defined(CONFIG_IDF_TARGET_ESP32C3) || \
+       defined(CONFIG_IDF_TARGET_ESP32C5) || \
+       defined(CONFIG_IDF_TARGET_ESP32C6) || \
+       defined(CONFIG_IDF_TARGET_ESP32H2) || \
+       defined(CONFIG_IDF_TARGET_ESP32P4))
+	esp_fill_random(input, RANDBYTES);
+#endif
+
 	/* Generate salt. */
 	workf = (factor < 4 || factor > 31)?12:factor;
 	aux = crypt_gensalt_rn("$2a$", workf, input, RANDBYTES,
diff --git a/crypt_blowfish/ow-crypt.h b/crypt_blowfish/ow-crypt.h
index 2e48794..721551e 100644
--- a/crypt_blowfish/ow-crypt.h
+++ b/crypt_blowfish/ow-crypt.h
@@ -17,27 +17,22 @@
 #ifndef _OW_CRYPT_H
 #define _OW_CRYPT_H
 
-#ifndef __GNUC__
-#undef __const
-#define __const const
-#endif
-
 #ifndef __SKIP_GNU
-extern char *crypt(__const char *key, __const char *setting);
-extern char *crypt_r(__const char *key, __const char *setting, void *data);
+extern char *crypt(const char *key, const char *setting);
+extern char *crypt_r(const char *key, const char *setting, void *data);
 #endif
 
 #ifndef __SKIP_OW
-extern char *crypt_rn(__const char *key, __const char *setting,
+extern char *crypt_rn(const char *key, const char *setting,
 	void *data, int size);
-extern char *crypt_ra(__const char *key, __const char *setting,
+extern char *crypt_ra(const char *key, const char *setting,
 	void **data, int *size);
-extern char *crypt_gensalt(__const char *prefix, unsigned long count,
-	__const char *input, int size);
-extern char *crypt_gensalt_rn(__const char *prefix, unsigned long count,
-	__const char *input, int size, char *output, int output_size);
-extern char *crypt_gensalt_ra(__const char *prefix, unsigned long count,
-	__const char *input, int size);
+extern char *crypt_gensalt(const char *prefix, unsigned long count,
+	const char *input, int size);
+extern char *crypt_gensalt_rn(const char *prefix, unsigned long count,
+	const char *input, int size, char *output, int output_size);
+extern char *crypt_gensalt_ra(const char *prefix, unsigned long count,
+	const char *input, int size);
 #endif
 
 #endif
diff --git a/crypt_blowfish/wrapper.c b/crypt_blowfish/wrapper.c
index 1e49c90..7b9de38 100644
--- a/crypt_blowfish/wrapper.c
+++ b/crypt_blowfish/wrapper.c
@@ -109,7 +109,7 @@ static char *_crypt_retval_magic(char *retval, const char *setting,
  * hashing algorithms use less than 128 bytes of the data area.
  */
 
-char *__crypt_rn(__const char *key, __const char *setting,
+char *__crypt_rn(const char *key, const char *setting,
 	void *data, int size)
 {
 	if (setting[0] == '$' && setting[1] == '2')
@@ -126,7 +126,7 @@ char *__crypt_rn(__const char *key, __const char *setting,
 	return NULL;
 }
 
-char *__crypt_ra(__const char *key, __const char *setting,
+char *__crypt_ra(const char *key, const char *setting,
 	void **data, int *size)
 {
 	if (setting[0] == '$' && setting[1] == '2') {
@@ -148,7 +148,7 @@ char *__crypt_ra(__const char *key, __const char *setting,
 	return __des_crypt_r(key, setting, (struct crypt_data *)*data);
 }
 
-char *__crypt_r(__const char *key, __const char *setting,
+char *__crypt_r(const char *key, const char *setting,
 	struct crypt_data *data)
 {
 	return _crypt_retval_magic(
@@ -156,7 +156,7 @@ char *__crypt_r(__const char *key, __const char *setting,
 		setting, (char *)data, sizeof(*data));
 }
 
-char *__crypt(__const char *key, __const char *setting)
+char *__crypt(const char *key, const char *setting)
 {
 	return _crypt_retval_magic(
 		__crypt_rn(key, setting, &_ufc_foobar, sizeof(_ufc_foobar)),
