diff --git a/CMakeLists.txt b/CMakeLists.txt
index b2a60a1..d084388 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.1)
+cmake_minimum_required(VERSION 3.16)
 
 project(hydrogen LANGUAGES C)
 
@@ -101,73 +101,78 @@ target_include_directories("${PROJECT_NAME}"
 target_compile_options("${PROJECT_NAME}" PRIVATE ${compile_options})
 
 # Installation
+option(ENABLE_INSTALL "Enable installation" ON)
+if(ENABLE_INSTALL)
+  set(targets_export_name "${PROJECT_NAME}-targets")
 
-set(targets_export_name "${PROJECT_NAME}-targets")
+  install(TARGETS "${PROJECT_NAME}"
+          EXPORT "${targets_export_name}"
+          LIBRARY DESTINATION "${CMAKE_INSTALL_LIBDIR}"
+          ARCHIVE DESTINATION "${CMAKE_INSTALL_LIBDIR}")
 
-install(TARGETS "${PROJECT_NAME}"
-        EXPORT "${targets_export_name}"
-        LIBRARY DESTINATION "${CMAKE_INSTALL_LIBDIR}"
-        ARCHIVE DESTINATION "${CMAKE_INSTALL_LIBDIR}")
+  install(FILES ${header_files} DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}")
 
-install(FILES ${header_files} DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}")
+  # CMake find_package() support
 
-# CMake find_package() support
+  set(install_config_dir "${CMAKE_INSTALL_DATADIR}/cmake/${PROJECT_NAME}")
 
-set(install_config_dir "${CMAKE_INSTALL_DATADIR}/cmake/${PROJECT_NAME}")
+  set(targets_export_file_name "${targets_export_name}.cmake")
+  set(targets_export_file "${PROJECT_BINARY_DIR}/${targets_export_file_name}")
 
-set(targets_export_file_name "${targets_export_name}.cmake")
-set(targets_export_file "${PROJECT_BINARY_DIR}/${targets_export_file_name}")
+  install(EXPORT "${targets_export_name}"
+          FILE "${targets_export_file_name}"
+          NAMESPACE "${PROJECT_NAME}::"
+          DESTINATION "${install_config_dir}")
 
-install(EXPORT "${targets_export_name}"
-        FILE "${targets_export_file_name}"
-        NAMESPACE "${PROJECT_NAME}::"
-        DESTINATION "${install_config_dir}")
+  set(config_file_name "${PROJECT_NAME}-config.cmake")
+  set(config_template_file "${PROJECT_SOURCE_DIR}/cmake/${config_file_name}.in")
+  set(config_file "${PROJECT_BINARY_DIR}/${config_file_name}")
 
-set(config_file_name "${PROJECT_NAME}-config.cmake")
-set(config_template_file "${PROJECT_SOURCE_DIR}/cmake/${config_file_name}.in")
-set(config_file "${PROJECT_BINARY_DIR}/${config_file_name}")
+  configure_package_config_file("${config_template_file}" "${config_file}"
+                                INSTALL_DESTINATION "${install_config_dir}")
 
-configure_package_config_file("${config_template_file}" "${config_file}"
-                              INSTALL_DESTINATION "${install_config_dir}")
+  install(FILES "${config_file}" DESTINATION "${install_config_dir}")
 
-install(FILES "${config_file}" DESTINATION "${install_config_dir}")
+  export(EXPORT "${targets_export_name}" FILE "${targets_export_file}" NAMESPACE "${PROJECT_NAME}::")
 
-export(EXPORT "${targets_export_name}" FILE "${targets_export_file}" NAMESPACE "${PROJECT_NAME}::")
+  export(PACKAGE "${PROJECT_NAME}")
+endif()
 
-export(PACKAGE "${PROJECT_NAME}")
 
 # Tests
-
-set(tests_executable "${PROJECT_NAME}-tests")
-set(tests_run_target "${PROJECT_NAME}-run-tests")
-set(tests_run_file "${PROJECT_BINARY_DIR}/${tests_run_target}.done")
-
-enable_testing()
-add_executable("${tests_executable}" ${test_files})
-target_compile_options("${tests_executable}" PRIVATE ${compile_options})
-target_link_libraries("${tests_executable}" "${PROJECT_NAME}")
-add_test(NAME "${tests_executable}" COMMAND "${tests_executable}")
-
-if(CMAKE_CROSSCOMPILING)
-  # Disable tests executable by default when cross-compiling (as it will fail to build when, e.g.,
-  # cross-compiling for Arduino/AVR).
-  set_target_properties("${tests_executable}"
-                        PROPERTIES
-                        EXCLUDE_FROM_ALL 1
-                        EXCLUDE_FROM_DEFAULT_BUILD 1)
-else()
-  # Otherwise, auto-run the tests on build.
-  add_custom_command(OUTPUT "${tests_run_file}"
-                     DEPENDS "${tests_executable}"
-                     COMMAND "${CMAKE_COMMAND}"
-                     ARGS -E remove "${tests_run_file}"
-                     COMMAND "${CMAKE_CTEST_COMMAND}"
-                     ARGS -C $<CONFIGURATION> --output-on-failure
-                     COMMAND "${CMAKE_COMMAND}"
-                     ARGS -E touch "${tests_run_file}"
-                     WORKING_DIRECTORY "${PROJECT_BINARY_DIR}"
-                     VERBATIM)
-  add_custom_target("${tests_run_target}" ALL DEPENDS "${tests_run_file}" VERBATIM)
+option(ENABLE_TESTS "Enable tests" OFF)
+if(ENABLE_TESTS)
+  set(tests_executable "${PROJECT_NAME}-tests")
+  set(tests_run_target "${PROJECT_NAME}-run-tests")
+  set(tests_run_file "${PROJECT_BINARY_DIR}/${tests_run_target}.done")
+
+  enable_testing()
+  add_executable("${tests_executable}" ${test_files})
+  target_compile_options("${tests_executable}" PRIVATE ${compile_options})
+  target_link_libraries("${tests_executable}" "${PROJECT_NAME}")
+  add_test(NAME "${tests_executable}" COMMAND "${tests_executable}")
+
+  if(CMAKE_CROSSCOMPILING)
+    # Disable tests executable by default when cross-compiling (as it will fail to build when, e.g.,
+    # cross-compiling for Arduino/AVR).
+    set_target_properties("${tests_executable}"
+                          PROPERTIES
+                          EXCLUDE_FROM_ALL 1
+                          EXCLUDE_FROM_DEFAULT_BUILD 1)
+  else()
+    # Otherwise, auto-run the tests on build.
+    add_custom_command(OUTPUT "${tests_run_file}"
+                      DEPENDS "${tests_executable}"
+                      COMMAND "${CMAKE_COMMAND}"
+                      ARGS -E remove "${tests_run_file}"
+                      COMMAND "${CMAKE_CTEST_COMMAND}"
+                      ARGS -C $<CONFIGURATION> --output-on-failure
+                      COMMAND "${CMAKE_COMMAND}"
+                      ARGS -E touch "${tests_run_file}"
+                      WORKING_DIRECTORY "${PROJECT_BINARY_DIR}"
+                      VERBATIM)
+    add_custom_target("${tests_run_target}" ALL DEPENDS "${tests_run_file}" VERBATIM)
+  endif()
 endif()
 
 # Generate Arduino package
diff --git a/impl/random.h b/impl/random.h
index bb18c5c..62c53b1 100644
--- a/impl/random.h
+++ b/impl/random.h
@@ -7,7 +7,16 @@ static TLS struct {
 
 #if defined(AVR) && !defined(__unix__)
 #    include "random/avr.h"
-#elif (defined(ESP32) || defined(ESP8266)) && !defined(__unix__)
+#elif (defined(CONFIG_IDF_TARGET_ESP32) || \
+       defined(ESP8266) || \
+       defined(CONFIG_IDF_TARGET_ESP32S2) || \
+       defined(CONFIG_IDF_TARGET_ESP32S3) || \
+       defined(CONFIG_IDF_TARGET_ESP32C2) || \
+       defined(CONFIG_IDF_TARGET_ESP32C3) || \
+       defined(CONFIG_IDF_TARGET_ESP32C5) || \
+       defined(CONFIG_IDF_TARGET_ESP32C6) || \
+       defined(CONFIG_IDF_TARGET_ESP32H2) || \
+       defined(CONFIG_IDF_TARGET_ESP32P4)) && !defined(__unix__)
 #    include "random/esp32.h"
 #elif defined(PARTICLE) && defined(PLATFORM_ID) && PLATFORM_ID > 2 && !defined(__unix__)
 #    include "random/particle.h"
diff --git a/impl/random/esp32.h b/impl/random/esp32.h
index 145ec61..4f75bb7 100644
--- a/impl/random/esp32.h
+++ b/impl/random/esp32.h
@@ -1,10 +1,21 @@
 // Important: RF *must* be activated on ESP board
 // https://techtutorialsx.com/2017/12/22/esp32-arduino-random-number-generation/
-#ifdef ESP32
-#    include <esp_system.h>
+#if (defined(CONFIG_IDF_TARGET_ESP32) || \
+     defined(ESP8266) || \
+     defined(CONFIG_IDF_TARGET_ESP32S2) || \
+     defined(CONFIG_IDF_TARGET_ESP32S3) || \
+     defined(CONFIG_IDF_TARGET_ESP32C2) || \
+     defined(CONFIG_IDF_TARGET_ESP32C3) || \
+     defined(CONFIG_IDF_TARGET_ESP32C5) || \
+     defined(CONFIG_IDF_TARGET_ESP32C6) || \
+     defined(CONFIG_IDF_TARGET_ESP32H2) || \
+     defined(CONFIG_IDF_TARGET_ESP32P4))
+#include <esp_random.h>
+#include <freertos/FreeRTOS.h>
+#include <freertos/timers.h>
 #endif
 
-#ifdef ARDUINO
+#if defined(ARDUINO)
 #    include <Arduino.h>
 #endif
 
@@ -20,7 +31,12 @@ hydro_random_init(void)
     while (ebits < 256) {
         uint32_t r = esp_random();
 
+#ifdef ARDUINO
         delay(10);
+#endif
+#ifdef ESP32
+        vTaskDelay(10);
+#endif
         hydro_hash_update(&st, (const uint32_t *) &r, sizeof r);
         ebits += 32;
     }
